Asterix/Optical_System_functions.py

```diff
@@ -1831,21 +1901,104 @@ def __init__(self, list_os, list_os_names):
         known_keywords = list(dict.fromkeys(known_keywords))
 
-        # we add no
         # We remove arguments we know are wrong
         if 'DMphase' in known_keywords:
             known_keywords.remove('DMphase')
+            # there is at least a DM, we add voltage_vector as an authorize kw
+
+            known_keywords.append('voltage_vector')
+            self.EF_through = self._control_testbed_with_voltages(
+                self.EF_through)
+
+        # to avoid mis-use we only use specific keywords.
         known_keywords.remove('kwargs')
 
         self.EF_through = _clean_EF_through(self.EF_through, known_keywords)
 
         #initialize the max and sum of PSFs for the normalization to contrast
         self.measure_normalization()
 
+    def voltage_to_phases(self, actu_vect, einstein_sum=False):
+        """ --------------------------------------------------
+        Generate the phase applied on one DM for a give vector of actuator amplitude
+        We decided to do it without matrix multiplication to save time because a
+        lot of the time we have lot of zeros in it
+
+        The phase is define at the reference wl and multiply by wl_ratio in DM.EF_through
+
+
+        Parameters:
+        ----------
+        actu_vect : flaot or 1D array
+                    values of the amplitudes for each actuator
+        einstein_sum : boolean. default false
+                        Use numpy Einstein sum to sum the pushact[i]*actu_vect[i]
+                        gives the same results as normal sum. Seems ot be faster for unique actuator
+                        but slower for more complex phases
+
+        Return:
+        ------
+            3D array
+            phases map for each DMS in the same unit as actu_vect * DM_pushact by order of light path
+
+        AUTHOR : Johan Mazoyer
+        -------------------------------------------------- """
+        DMphases = np.zeros(
+            (self.number_DMs, self.dim_overpad_pupil, self.dim_overpad_pupil))
+        indice_acum_number_act = 0
+
+        if isinstance(actu_vect, float):
+            return np.zeros(self.number_DMs) + actu_vect
+
+        if len(actu_vect) != self.number_act:
+            raise Exception(
+                "voltage vector must be 0 or array of dimension testbed.number_act,"
+                + "sum of all DM.number_act")
+
+        for i, DM_name in enumerate(self.name_of_DMs):
+
+            DM = vars(self)[DM_name]
+            actu_vect_DM = actu_vect[
+                indice_acum_number_act:indice_acum_number_act + DM.number_act]
+            DMphases[i] = DM.voltage_to_phase(actu_vect_DM,
+                                              einstein_sum=einstein_sum)
+
+            indice_acum_number_act += DM.number_act
+
+        return DMphases
+
+    def _control_testbed_with_voltages(self, testbed_EF_through):
+        """ --------------------------------------------------
+        A function to controle the testbed with a single function
```

Comment:
I do not understand what this sentence means

typo : control instead of controle

Reply:
I corrected by 

A function to go from a testbed_EF_through with several DMXX_phase
        parameters (one for each DM), to a testbed_EF_through with a unique
        voltage_vector parameter of size testbed.number_act (or a single float, like 0.)

        the problem with DMXX_phase parameters is that it cannot be automated since it requires
        to know the name/number of the DMs in advance.

        DMXX_phase parameters can still be used, but are overridden by voltage_vector parameter
        if present.